### PR TITLE
Visualisation of the image & graph annotations

### LINKED ISSUE #108: [BUG] Napari binarises synthetic images
**Describe the bug**

I've created a dataset full of synthetic images where the entire image is set to value = 'grey' (127 for integer `dtype`, 0.5 for float `dtype`) and the squares under each node are set to the corresponding value: 'black' (0) for true negative nodes and 'white' (255 or 1.0) for true positive, object belonging nodes. I save these images as `mrc` files & check that their values are correct. 

In the napari annotator, the images are visualised as binary, i.e. only two colours in the black-and-white masks. Changing the 
``` 
with mrcfile.open(IMAGE_PATH, "r") as mrc:
    # image_data = mrc.data.astype(int)
    image_data = mrc.data 
```
^^^ Switching the `astype(int)` off or changing it to `astype(float)` doesn't help. Why is the annotator binarising the images? How can I turn this off?

image
Screenshot 2023-06-20 at 11 18 59

### LINKED ISSUE #136: Adding visualisation of graph to grace as an optional plot out
Since napari is not a part of installation , it would be good to plot out a visualisation of the annotation as a sanity check 

```fig, ax = plt.subplots(figsize=(16, 16))

# node positions
pos = {
    idx: (node[GraphAttrs.NODE_X], node[GraphAttrs.NODE_Y]) 
    for idx, node in graph.nodes(data=True)
}
# edge annotations
edge_gt = [graph[u][v][GraphAttrs.EDGE_GROUND_TRUTH] for u, v in graph.edges]
edge_colors = [EdgeColor[gt.name].value for gt in edge_gt]

node_colors = [
    EdgeColor[node_attrs[GraphAttrs.NODE_GROUND_TRUTH].name].value 
    for _, node_attrs in graph.nodes(data=True)
]

ax.imshow(dataset.annotation, cmap=plt.cm.turbo, interpolation="none")

# draw all nodes/vertices in the graph, including those not determined to be 
# part of the objects
nx.draw_networkx(
    dataset.graph, 
    ax=ax, 
    pos=pos, 
    with_labels=False, 
    # node_color="w", 
    node_size=32,
    edge_color=edge_colors,
    node_color=node_colors,
)

ax.set_title(dataset.metadata["image_filename"])```

### LINKED ISSUE #158: Keep edge attributes by removing edges
Currently, the GT graph is generated by copying the nodes from the original graph and connecting those nodes which are connected by a true positive (TP) edge. This maintains the node features but all edge features are lost.

### PR DESCRIPTION
**Pull request by KristinaUlicna**
When reviewing, please annotate an image in *napari* & then visualise it using the `read_grace.ipynb` notebook

**Data processing:**
- [x] Comments out the inflammatory `astype(int)` from the image processing. Mentioned in #148 
- [x] When annotating in napari, marks the protruding nodes forming an TN edge as TN node. Discussed in #143 
- [x] When annotating in napari, sets the contrast range of the image to min / max value. Resolves #108

**Plotting / visualisation:**
- [x] Visualise the grace annotations overlaid on the image.
- [x] Display a montage & average image per classifier training class.
- [x] Plots simple graph & identifies isolated objects (i.e. connected components)

**Graph analysis:**
- [x] Generates a ground truth (GT) graph to allow metric calculation
- [x] Maintains all node & edge attributes in new GT graph: resolves #158 
- [x] Shortlists visualisation options in `read_grace.ipynb` notebook: resolves #136

### CODE REVIEW

**Comment by KristinaUlicna** [comment]
Apologies for a seemingly massive PR, but all the changes are related to the visualisation / annotation update. 
This PR needs to go in before metrics can be evaluated. Happy to discuss 🚀

File: grace/evaluation/visualisation.py [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+
+def plot_simple_graph(G: nx.Graph, title: str = "") -> None:
+    """Plots a simple graph with black nodes and edges."""
+
+    # Fancy annotation plot
+    _, ax = plt.subplots(figsize=(16, 16))
+
+    # node positions
+    pos = {
+        idx: (node[GraphAttrs.NODE_X], node[GraphAttrs.NODE_Y])
```
Comment (mooniean) [inline grace/evaluation/visualisation.py:22]: Graphs have the X and the Y swapped, so when plotting this, the graph is transposed to the original image

File: grace/evaluation/visualisation.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+import networkx as nx
+import numpy as np
+
+import numpy.typing as npt
+
+from skimage.util import montage
+
+
+def plot_simple_graph(G: nx.Graph, title: str = "") -> None:
+    """Plots a simple graph with black nodes and edges."""
```
Comment (mooniean) [inline grace/evaluation/visualisation.py:15]: Add function input parameters

File: grace/evaluation/visualisation.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+        edge_color="k",
+        node_color="k",
+    )
+
+    ax.set_title(f"{title}")
+    plt.show()
+
+
+def plot_connected_components(G: nx.Graph, title: str = "") -> None:
+    """Colour-codes the connected components (individual objects)
```
Comment (mooniean) [inline grace/evaluation/visualisation.py:42]: same as above - add input parameters

**Review by mooniean — COMMENTED** [review]
made some comments, but everything was running!

File: grace/evaluation/visualisation.py (hunk 2 above)
Comment (KristinaUlicna) [reply grace/evaluation/visualisation.py:15]: Input params are typed in directly, there's not much more to describe here as their names are self-explanatory.

File: grace/evaluation/visualisation.py (hunk 3 above)
Comment (KristinaUlicna) [reply grace/evaluation/visualisation.py:42]: Just a helper fn to visualise the graph, not much to explain.

**Comment by KristinaUlicna** [comment]
> made some comments, but everything was running!

@mooniean Could you comment on what you did / run to review? It's a significant PR with important data processing, I'd like to make sure it was checked correctly.

**Comment by mooniean** [comment]
> > made some comments, but everything was running!
> 
> @mooniean Could you comment on what you did / run to review? It's a significant PR with important data processing, I'd like to make sure it was checked correctly.

I've done the requested steps: opened image, annotated in napari, loaded it with ```read_grace.ipynb``` and ran the whole notebook. Everything was working, the correct things were being shown (apart from the graph being transposed without the image)

File: grace/evaluation/visualisation.py (hunk 1 above)
Comment (KristinaUlicna) [reply grace/evaluation/visualisation.py:22]: This will get fixed in the subsequent PR.

**Review by mooniean — APPROVED** [review]

**Pull request merged**